Constant vector source

amcastro-tri opened the pull request:
This PR implements a constant source system as discussed in #2986.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3116)
<!-- Reviewable:end -->

amcastro-tri commented:
The usual suspects +@david-german-tri and +@sherm1 for feature review please.
+@jwnimmer-tri for platform please.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-:-KOu5a68pPTQ0JsEVZX-:b-nb4x9j)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review pass done, PTAL.

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/constant_source.cc, line 80 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuIMPIUF1JToW1RPFZ:-KOuIMPIUF1JToW1RPF_:b-qktjcf) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L80)):*
> ```cpp
> }
> 
> template class DRAKESYSTEMFRAMEWORK_EXPORT ConstantSource<double>;
> ```

Needs a comment saying why this is here. Maybe: "// Pre-instantiate most common variant and verify that it compiles successfully."

---

*[drake/systems/framework/primitives/constant_source.h, line 19 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuH1wjz_iFna1VbTxA:-KOuH1wk-or3_wFVlzwP:b-cpjvlh) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.h#L19)):*
> ```objc
> /// @tparam T The vector element type, which must be a valid Eigen scalar.
> template <typename T>
> class ConstantSource : public System<T> {
> ```

This should be named `ConstantVectorSource` since it only supports vector values. Save `ConstantSource` for a block that can provide abstract values (values of any type).

---

*[drake/systems/framework/primitives/constant_source.h, line 21 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuGYoe3kirM2L35O7R:-KOuGYoe3kirM2L35O7S:b-4c6dru) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.h#L21)):*
> ```objc
> class ConstantSource : public System<T> {
>  public:
>   /// @param source_value the constant value of the output so that
> ```

Doxygen needs a first sentence that it can use as a brief for this method.

---

*[drake/systems/framework/primitives/constant_source.h, line 33 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuGozdPztTBlHBbSNl:-KOuGozdPztTBlHBbSNm:bd6k5hw) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.h#L33)):*
> ```objc
>       const ContextBase<T>& context) const override;
> 
>   /// Outputs a signal with a fixed value as specified by the user.
> ```

BTW, this is a doxygen comment but the previous two were developer comments -- is that what you wanted?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-:-KOuJMmXyyxpS3qO2SNv:b-l9l3kz)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/constant_source.cc, line 3 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMLP4Sl6unSgMhLsz:-KOuMLP4Sl6unSgMhLt-:b-wxmpju) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L3)):*
> ```cpp
> #include "drake/systems/framework/primitives/constant_source.h"
> 
> #include <stdexcept>
> ```

Unused?

---

*[drake/systems/framework/primitives/constant_source.cc, line 4 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMMGWnJCgnLpIYydY:-KOuMMGWnJCgnLpIYydZ:b-ciom1o) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L4)):*
> ```cpp
> 
> #include <stdexcept>
> #include <string>
> ```

Unused?

---

*[drake/systems/framework/primitives/constant_source.cc, line 19 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMbpnhROYVPALat00:-KOuMbpoehluJRPQlDGK:b-yas518) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L19)):*
> ```cpp
>     source_value_(source_value) {
>   // Input source must have at least one entry.
>   DRAKE_ABORT_UNLESS(source_value.rows() > 0);
> ```

The API does not warn the calling code that this requirement exists.

---

*[drake/systems/framework/primitives/constant_source.cc, line 28 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMq1sNt6NY9cXm2rf:-KOuMq1sNt6NY9cXm2rg:bats7wc) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L28)):*
> ```cpp
> ConstantSource<T>::CreateDefaultContext() const {
>   std::unique_ptr<Context<T>> context(new Context<T>);
>   // TODO(amcastro-tri): Remove this override after #3102 is merged.
> ```

What override?

---

*[drake/systems/framework/primitives/constant_source.cc, line 38 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMuEjWK1Gzoy504sK:-KOuMuEjWK1Gzoy504sL:bats7wc) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L38)):*
> ```cpp
> std::unique_ptr<SystemOutput<T>> ConstantSource<T>::AllocateOutput(
>     const ContextBase<T>& context) const {
>   // TODO(amcastro-tri): Remove this override after #3102 is merged.
> ```

What override?

---

*[drake/systems/framework/primitives/constant_source.cc, line 54 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuN-i-3fXBHDkdI97H:-KOuN-i0PAPG17ERt3k2:b-l1v198) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L54)):*
> ```cpp
> template <typename T>
> void ConstantSource<T>::EvalOutput(const ContextBase<T>& context,
>                           SystemOutput<T>* output) const {
> ```

Indentation seems weird.

---

*[drake/systems/framework/primitives/constant_source.h, line 7 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMBIbGU5A4ZqnA2Qz:-KOuMBIcTGORPPyCqgyg:b-ciom1o) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.h#L7)):*
> ```objc
> 
> #include "drake/common/eigen_types.h"
> #include "drake/systems/framework/cache.h"
> ```

Unused?

---

*[drake/systems/framework/primitives/constant_source.h, line 11 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuM0tWW2a31qhLc14V:-KOuM0tWW2a31qhLc14W:b-ciom1o) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.h#L11)):*
> ```objc
> #include "drake/systems/framework/system.h"
> #include "drake/systems/framework/system_output.h"
> #include "drake/systems/framework/vector_interface.h"
> ```

Unused?

---

*[drake/systems/framework/primitives/test/constant_source_test.cc, line 35 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuTbnzZDbzlt0BvkWi:-KOuTbo-6wP-hdA5uYrM:bo5copq) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_source_test.cc#L35)):*
> ```cpp
>   }
> 
>   const Vector2d kConstantSource_{2.0, 1.5};
> ```

This likely needs Eigen alignment mumbo-jumbo?

---

*[drake/systems/framework/primitives/test/constant_source_test.cc, line 35 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuUsdl6zZ0wKABHEAp:-KOuUsdl6zZ0wKABHEAq:bs4v01h) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_source_test.cc#L35)):*
> ```cpp
>   }
> 
>   const Vector2d kConstantSource_{2.0, 1.5};
> ```

It seems quite unconventional to me to have a `kConstant_` with the trailing underbar.  If you are wed to this notation, I can go check the cppguide to see if its allowed, but perhaps it's worth changing no matter what, so as not to be weird?

---

*[drake/systems/framework/primitives/test/constant_source_test.cc, line 44 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuTwlsnY3eyG2MzKrk:-KOuTwlsnY3eyG2MzKrl:b-eypuzo) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_source_test.cc#L44)):*
> ```cpp
> TEST_F(ConstantSourceTest, OutputTest) {
>   // TODO(amcastro-tri): we should be able to ask:
>   // source_->num_of_input_ports().
> ```

BTW I am unclear if the dozen or so TODOs hanging around everywhere in this PR are "everyone has mostly agreed on the proposed change in the TODO, but we haven't gotten around to implementing it yet" or "I wish it were this way, and I'm going to use TODOs on master to share my design".  If the former, that's great; I am skeptical of TODOs as a technique to share the latter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/framework/primitives/test/constant_source_test.cc, line 35 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuTbnzZDbzlt0BvkWi:-KOusqWYwSoobckj5rtW:b-pmshm8) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_source_test.cc#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This likely needs Eigen alignment mumbo-jumbo?
</details>
FYI a non-footgun version of Point2<double>:
https://github.com/RobotLocomotion/drake/blob/8a0b51c720f0f1e0bf946128fb5d5d7dbab82ac3/drake/examples/Cars/curve2.h#L23

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I just implemented the auto-diff unit test using @jwnimmer-tri's explicit instantiation framework using the `-inl.h` files. PTAL.

---

Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-:-KOuztT4c-38uQodpFx7:b9koekv)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Sorry, disregard the last comment here. Wrong PR.

---

Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-:-KOv-CzxCGdVxcTUSw31:b-lrwo1y)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[drake/systems/framework/primitives/constant_source.cc, line 19 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMbpnhROYVPALat00:-KOx2YhS9ScCSMicdDeu:b-5y98bc) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L19)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The API does not warn the calling code that this requirement exists.
</details>
BTW, not clear this really should be a requirement. Is there any reason these things can't work correctly with 0-length vectors? Because a system diagram can represent a very general computation it is possible for 0-length outputs to occur. For example, consider a block that selects only the positive elements of an input vector -- there might not be any. 0-length vectors and matrices come up when slicing and extracting matrix blocks; I believe they work fine in Eigen.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 3 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 7 files reviewed at latest revision, 15 unresolved discussions.

---

*[drake/systems/framework/primitives/test/constant_vector_source_test.cc, line 44 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuTwlsnY3eyG2MzKrk:-KP8eG87wAOOY47vAy-U:b-bqt14m) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_vector_source_test.cc#L44)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I am unclear if the dozen or so TODOs hanging around everywhere in this PR are "everyone has mostly agreed on the proposed change in the TODO, but we haven't gotten around to implementing it yet" or "I wish it were this way, and I'm going to use TODOs on master to share my design".  If the former, that's great; I am skeptical of TODOs as a technique to share the latter.
</details>
Well... actually I think a mixture of both. As one of the first users of System 2.0 I am helping @david-german-tri (or trying) on identifying pain points for the user. Many of these TODO don't necessarily need to make it to master but are meant to initiate discussion with @david-german-tri and @sherm1. Many of them should have an issue (for instance I just filed #3140) and for many of them @david-german-tri himself either files and issue or a PR (see #3102). As @david-german-tri put it on Friday, System 2.0 is actively under development right now. Probably here the list of TODO's is border line annoying but in other cases this approach helped identify real issues (see  #3132).

---

*[drake/systems/framework/primitives/constant_source.cc, line 3 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMLP4Sl6unSgMhLsz:-KP8TtADa84OYSgF_EoA:b-hfpv0l) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L3)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Unused?
</details>
removed

---

*[drake/systems/framework/primitives/constant_source.cc, line 4 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMMGWnJCgnLpIYydY:-KP8Tu0hMJeeZcvfOjSi:b-hfpv0l) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L4)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Unused?
</details>
removed

---

*[drake/systems/framework/primitives/constant_source.cc, line 19 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMbpnhROYVPALat00:-KP8UI8H4gxuWotILfN9:bfcfdsv) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L19)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, not clear this really should be a requirement. Is there any reason these things can't work correctly with 0-length vectors? Because a system diagram can represent a very general computation it is possible for 0-length outputs to occur. For example, consider a block that selects only the positive elements of an input vector -- there might not be any. 0-length vectors and matrices come up when slicing and extracting matrix blocks; I believe they work fine in Eigen.
</details>
Good point @sherm1. I removed the assert. @jwnimmer-tri, with the assert removed now there is no need to change the API doc. 

---

*[drake/systems/framework/primitives/constant_source.cc, line 28 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMq1sNt6NY9cXm2rf:-KP8V0wvN-WJTA_yPR8w:b29h69k) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> What override?
</details>
paraphrased. 

---

*[drake/systems/framework/primitives/constant_source.cc, line 38 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMuEjWK1Gzoy504sK:-KP8V6fOcZ7gWjkSE2hf:b-93bkyg) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L38)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> What override?
</details>
paraphrased.

---

*[drake/systems/framework/primitives/constant_source.cc, line 54 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuN-i-3fXBHDkdI97H:-KP8VHIuG3dh76DB_zcd:b56yn6i) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L54)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Indentation seems weird.
</details>
good catch. Thank you

---

*[drake/systems/framework/primitives/constant_source.cc, line 80 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuIMPIUF1JToW1RPFZ:-KP8Vi399gTaZtMiNm93:bfn5czz) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.cc#L80)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Needs a comment saying why this is here. Maybe: "// Pre-instantiate most common variant and verify that it compiles successfully."
</details>
Done. Also added same line to `adder.cc` and `integrator.cc`.

---

*[drake/systems/framework/primitives/constant_source.h, line 7 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuMBIbGU5A4ZqnA2Qz:-KP8WAMToRpDqdXKxrNf:b-hfpv0l) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.h#L7)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Unused?
</details>
removed

---

*[drake/systems/framework/primitives/constant_source.h, line 11 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuM0tWW2a31qhLc14V:-KP8WAvvrhtolMuMIuib:b-hfpv0l) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.h#L11)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Unused?
</details>
removed

---

*[drake/systems/framework/primitives/constant_source.h, line 19 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuH1wjz_iFna1VbTxA:-KP8YFEX1ULVNecJifPh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.h#L19)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This should be named `ConstantVectorSource` since it only supports vector values. Save `ConstantSource` for a block that can provide abstract values (values of any type).
</details>
Done.

---

*[drake/systems/framework/primitives/constant_source.h, line 21 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuGYoe3kirM2L35O7R:-KP8YkuarsirCYrt1v7W:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.h#L21)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Doxygen needs a first sentence that it can use as a brief for this method.
</details>
Done.

---

*[drake/systems/framework/primitives/constant_source.h, line 33 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuGozdPztTBlHBbSNl:-KP8YsaGHRoo1hYjLKrM:b9ccc3o) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_source.h#L33)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, this is a doxygen comment but the previous two were developer comments -- is that what you wanted?
</details>
Yes. `EvalOutput` can actually be called by the user while the previous two will have a default implementation as soon as #3102 lands.

---

*[drake/systems/framework/primitives/test/constant_source_test.cc, line 35 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuTbnzZDbzlt0BvkWi:-KP8ZIJBiAIufXQZQfWn:b-ogfjan) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_source_test.cc#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> FYI a non-footgun version of Point2<double>:
> https://github.com/RobotLocomotion/drake/blob/8a0b51c720f0f1e0bf946128fb5d5d7dbab82ac3/drake/examples/Cars/curve2.h#L23
</details>
Why would I turn off alignment?

---

*[drake/systems/framework/primitives/test/constant_source_test.cc, line 35 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuUsdl6zZ0wKABHEAp:-KP8_lso7g_T2cyV3l_G:b-bh6kp8) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_source_test.cc#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> It seems quite unconventional to me to have a `kConstant_` with the trailing underbar.  If you are wed to this notation, I can go check the cppguide to see if its allowed, but perhaps it's worth changing no matter what, so as not to be weird?
</details>
I didn't see much in the style guide. However I double checked and in our code base it seems we are opting for no trailing `_`. So I removed it to be consistent. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 10 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-:-KP9W0Trk1zxOusYq9j_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 10 of 10 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/framework/primitives/constant_vector_source.cc, line 54 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuN-i-3fXBHDkdI97H:-KP9gwPYK7mPn1q84WGH:b-f5bnus) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_vector_source.cc#L54)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> good catch. Thank you
</details>
I see no changes here?

---

*[drake/systems/framework/primitives/constant_vector_source.h, line 33 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuGozdPztTBlHBbSNl:-KP9hPpMaEPJl1u_aTTG:b16ygbp) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_vector_source.h#L33)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Yes. `EvalOutput` can actually be called by the user while the previous two will have a default implementation as soon as #3102 lands.
</details>
The comment style should reflect the code as it stands now, not as how it will be.  They are all overrides, so it seems to me that they should either all be doxygen or all be non-doxygen.  Given the prevailing System 2.0 style (which seems to be to re-comment overrides in doxygen style, with their specific behavior), I think these should all be doxygen, right?

---

*[drake/systems/framework/primitives/test/constant_vector_source_test.cc, line 35 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuTbnzZDbzlt0BvkWi:-KP9hfMs2fQXCtEtgto_:baszr00) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_vector_source_test.cc#L35)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Why would I turn off alignment?
</details>
My hypothesis was that by having a fixed-size Eigen member field, you'd need `EIGEN_MAKE_ALIGNED_OPERATOR_NEW` goop on this class.  By marking the object here as non-aligned, you wouldn't need it.  But if you don't need the goop, then you don't need the alignment-disable either.

In any case, Eigen manual tells me that structures having Vector2d has a member field need alignment magic, so I think it's still a defect that the alignment here is un-addressed in any way.

---

*[drake/systems/framework/primitives/test/constant_vector_source_test.cc, line 44 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuTwlsnY3eyG2MzKrk:-KP9hEcmqITSnEvcM_7G:bok9zw2) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_vector_source_test.cc#L44)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Well... actually I think a mixture of both. As one of the first users of System 2.0 I am helping @david-german-tri (or trying) on identifying pain points for the user. Many of these TODO don't necessarily need to make it to master but are meant to initiate discussion with @david-german-tri and @sherm1. Many of them should have an issue (for instance I just filed #3140) and for many of them @david-german-tri himself either files and issue or a PR (see #3102). As @david-german-tri put it on Friday, System 2.0 is actively under development right now. Probably here the list of TODO's is border line annoying but in other cases this approach helped identify real issues (see  #3132).
</details>
My point is not whether they are annoying; by agreeing to a PR that adds them to master, we are agreeing that they should be done.  If they are just margin notes, don't put them into master.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 1 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/framework/primitives/constant_vector_source.cc, line 54 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuN-i-3fXBHDkdI97H:-KPIWfUCR_XgTHmyUrGe:binjo3j) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_vector_source.cc#L54)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I see no changes here?
</details>
sorry about that. I was so sure I had made the change. Done now

---

*[drake/systems/framework/primitives/constant_vector_source.h, line 33 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuGozdPztTBlHBbSNl:-KPI_I-fGh2pxpIRTqRw:bs2ghjp) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_vector_source.h#L33)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The comment style should reflect the code as it stands now, not as how it will be.  They are all overrides, so it seems to me that they should either all be doxygen or all be non-doxygen.  Given the prevailing System 2.0 style (which seems to be to re-comment overrides in doxygen style, with their specific behavior), I think these should all be doxygen, right?
</details>
#3102 landed so now we have less methods to implement! (and document) hurray! 

---

*[drake/systems/framework/primitives/test/constant_vector_source_test.cc, line 35 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuTbnzZDbzlt0BvkWi:-KPIa1jE3gxaaPYPjvee:bbcdjbu) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_vector_source_test.cc#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> My hypothesis was that by having a fixed-size Eigen member field, you'd need `EIGEN_MAKE_ALIGNED_OPERATOR_NEW` goop on this class.  By marking the object here as non-aligned, you wouldn't need it.  But if you don't need the goop, then you don't need the alignment-disable either.
> 
> In any case, Eigen manual tells me that structures having Vector2d has a member field need alignment magic, so I think it's still a defect that the alignment here is un-addressed in any way.
</details>
Correct. I had forgotten when reading your comment this was in a fixture. I used your suggestion with `Eigen::DontAlign`.

---

*[drake/systems/framework/primitives/test/constant_vector_source_test.cc, line 44 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuTwlsnY3eyG2MzKrk:-KPIaKEvY1wJ7e_94w6P:buvr3wv) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_vector_source_test.cc#L44)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> My point is not whether they are annoying; by agreeing to a PR that adds them to master, we are agreeing that they should be done.  If they are just margin notes, don't put them into master.
</details>
I understand your concern. We discussed many of these with @david-german-tri and I removed a lot of them after #3102. The only ones that stayed have an issue to solve them (like #3140 to be pushed today or PR #3097 under review right now).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 10 files at r3, 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/framework/CMakeLists.txt, line 46 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KPIex6K6v9spL-MXPKl:-KPIex6K6v9spL-MXPKm:blbhhbe) ([raw file](https://github.com/robotlocomotion/drake/blob/cdfd0729b93a847b8a51e9dc0df595cb8724f63b/drake/systems/framework/CMakeLists.txt#L46)):*
> ```CMake
>   primitives/adder.h
>   primitives/constant_vector_source.h
>   primitives/integrator.h
> ```

You can't have `/` in `drake_install_headers` targets -- the `primitives/` part gets ignored by CMake.  If you look on disk, all of these primitives are being installed into e.g. `build/install/include/drake/systems/framework/adder.h`, which is not what you want.  TTBOMK, you need to add a `primitives/CMakeLists.txt` that installs these headers; there isn't a way to do it from this file.  (Perhaps the list of subdirectory headers to install should still be populated here, and the subdir logic just has the install command on that variable, so that the lists are easier to keep in sync?)

/CC @david-german-tri

---

*[drake/systems/framework/primitives/constant_vector_source.h, line 7 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KPIfmVj1aNitCI93cyX:-KPIfmVj1aNitCI93cyY:b-numljj) ([raw file](https://github.com/robotlocomotion/drake/blob/cdfd0729b93a847b8a51e9dc0df595cb8724f63b/drake/systems/framework/primitives/constant_vector_source.h#L7)):*
> ```objc
> 
> #include "drake/common/eigen_types.h"
> ```

Spurious blank line.

---

*[drake/systems/framework/primitives/test/constant_vector_source_test.cc, line 44 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuTwlsnY3eyG2MzKrk:-KPIecBLY5Q3GvLOGN4m:bszelpd) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_vector_source_test.cc#L44)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I understand your concern. We discussed many of these with @david-german-tri and I removed a lot of them after #3102. The only ones that stayed have an issue to solve them (like #3140 to be pushed today or PR #3097 under review right now).
</details>
Great, thank you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/framework/CMakeLists.txt, line 46 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KPIex6K6v9spL-MXPKl:-KPIyorMl71ECMibSmpq:b-cklwhg) ([raw file](https://github.com/robotlocomotion/drake/blob/cdfd0729b93a847b8a51e9dc0df595cb8724f63b/drake/systems/framework/CMakeLists.txt#L46)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> You can't have `/` in `drake_install_headers` targets -- the `primitives/` part gets ignored by CMake.  If you look on disk, all of these primitives are being installed into e.g. `build/install/include/drake/systems/framework/adder.h`, which is not what you want.  TTBOMK, you need to add a `primitives/CMakeLists.txt` that installs these headers; there isn't a way to do it from this file.  (Perhaps the list of subdirectory headers to install should still be populated here, and the subdir logic just has the install command on that variable, so that the lists are easier to keep in sync?)
> 
> /CC @david-german-tri
</details>
Thank you for your comment @jwnimmer-tri, I was following @david-german-tri's pattern here.
@david-german-tri, is there a reason why we are including these this way? 

---

*[drake/systems/framework/primitives/constant_vector_source.h, line 7 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KPIfmVj1aNitCI93cyX:-KPIzw3cNhEoLyGc7EIg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cdfd0729b93a847b8a51e9dc0df595cb8724f63b/drake/systems/framework/primitives/constant_vector_source.h#L7)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Spurious blank line.
</details>
Done.

---

*[drake/systems/framework/primitives/test/constant_vector_source_test.cc, line 44 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuTwlsnY3eyG2MzKrk:-KPIzzIg-BxTSwAf0y8X:bsv417h) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/test/constant_vector_source_test.cc#L44)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Great, thank you.
</details>
thank you!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 10 files at r3, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/CMakeLists.txt, line 46 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KPIex6K6v9spL-MXPKl:-KPJ-KV72sPxUq9DglHu:b-3fv7bu) ([raw file](https://github.com/robotlocomotion/drake/blob/cdfd0729b93a847b8a51e9dc0df595cb8724f63b/drake/systems/framework/CMakeLists.txt#L46)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Thank you for your comment @jwnimmer-tri, I was following @david-german-tri's pattern here.
> @david-german-tri, is there a reason why we are including these this way? 
</details>
I think it's just been wrong since forever (in Cars, I had this wrong for weeks as well; it's a non-obvious failure mode).  I'd be OK if you all preferred a follow-up PR.  I will mark myself as satisfied here so you can decide.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-:-KPJ-YYktldoOMSM0DvR:bnfp4nl)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/CMakeLists.txt, line 46 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KPIex6K6v9spL-MXPKl:-KPJ-c1ly7b3oWXhsnKJ:bx86ud8) ([raw file](https://github.com/robotlocomotion/drake/blob/cdfd0729b93a847b8a51e9dc0df595cb8724f63b/drake/systems/framework/CMakeLists.txt#L46)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I think it's just been wrong since forever (in Cars, I had this wrong for weeks as well; it's a non-obvious failure mode).  I'd be OK if you all preferred a follow-up PR.  I will mark myself as satisfied here so you can decide.
</details>
Yes, it's been wrong forever.  I agree it should be cleaned up in a separate PR; I'll do that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/framework/CMakeLists.txt, line 46 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KPIex6K6v9spL-MXPKl:-KPJ3SOyl0D8dJSHuhhh:b-7w9igh) ([raw file](https://github.com/robotlocomotion/drake/blob/cdfd0729b93a847b8a51e9dc0df595cb8724f63b/drake/systems/framework/CMakeLists.txt#L46)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Yes, it's been wrong forever.  I agree it should be cleaned up in a separate PR; I'll do that.
</details>
@david-german-tri, this one is blocked on you. Should I wait for your fix?

---

*[drake/systems/framework/primitives/constant_vector_source.h, line 33 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuGozdPztTBlHBbSNl:-KPJ3Ke6sHZ04MA6x209:bpc89sd) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_vector_source.h#L33)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> #3102 landed so now we have less methods to implement! (and document) hurray! 
</details>
@sherm1, this one is blocked on you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/framework/CMakeLists.txt, line 46 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KPIex6K6v9spL-MXPKl:-KPJ4FxfvIbFm91PRn99:b-v3bmtq) ([raw file](https://github.com/robotlocomotion/drake/blob/cdfd0729b93a847b8a51e9dc0df595cb8724f63b/drake/systems/framework/CMakeLists.txt#L46)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> @david-german-tri, this one is blocked on you. Should I wait for your fix?
</details>
unblocked, sorry.  No need to wait.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[drake/systems/framework/primitives/constant_vector_source.h, line 33 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-KOuGozdPztTBlHBbSNl:-KPJCkHhCzIbltr9cTcV:br4vygh) ([raw file](https://github.com/robotlocomotion/drake/blob/ad19fb55ba57c6f8310bee294a62d7e2cbbeac4c/drake/systems/framework/primitives/constant_vector_source.h#L33)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> @sherm1, this one is blocked on you.
</details>
Since Sherm used BTW, it seems safe to unblock the thread.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3116#-:-KPJCzbBWCYl_kOzuzem:bnfp4nl)*
<!-- Sent from Reviewable.io -->
